app_flutter/lib/widgets/lattice.dart

```diff
@@ -0,0 +1,876 @@
+@immutable
+class LatticeCell extends _LatticeCell {
+  const LatticeCell({
+    Painter painter,
+    this.builder,
+    LatticeTapCallback onTap,
+  }) : super(painter: painter, onTap: onTap);
+
+  final WidgetBuilder builder;
+
+  @override
+  bool get hasChild => builder != null;
+}
+
+/// A bidirectional scrollable view that draws arrays of arrays of [LatticeCell]s.
+///
+/// Only the [cells] that are visible are drawn.
+///
+/// The cells will be sized according to [cellSize].
+class LatticeScrollView extends StatelessWidget {
+  const LatticeScrollView({
+    Key key,
+    this.horizontalPhysics,
+    this.horizontalController,
+    this.textDirection,
+    this.verticalPhysics,
+    this.verticalController,
+    this.dragStartBehavior = DragStartBehavior.start,
+    @required this.cells,
+    @required this.cellSize,
+  })  : assert(dragStartBehavior != null),
+        assert(cellSize != null),
+        assert(cells != null),
+        super(key: key);
+
+  final ScrollPhysics horizontalPhysics;
+
+  final ScrollController horizontalController;
+
+  final TextDirection textDirection;
+
+  final ScrollPhysics verticalPhysics;
+
+  final ScrollController verticalController;
+
+  final DragStartBehavior dragStartBehavior;
+
+  final List<List<LatticeCell>> cells;
+
+  final Size cellSize;
+
+  @override
+  Widget build(BuildContext context) {
+    final TextDirection textDirection = this.textDirection ?? Directionality.of(context);
+    return Scrollable(
+      dragStartBehavior: dragStartBehavior,
+      axisDirection: textDirectionToAxisDirection(textDirection),
+      controller: horizontalController,
+      physics: horizontalPhysics,
+      viewportBuilder: (BuildContext context, ViewportOffset horizontalOffset) => _FakeViewport(
+        child: Scrollable(
+          dragStartBehavior: dragStartBehavior,
+          axisDirection: AxisDirection.down,
+          controller: verticalController,
+          physics: verticalPhysics,
+          viewportBuilder: (BuildContext context, ViewportOffset verticalOffset) => _FakeViewport(
+            child: _LatticeBody(
+              textDirection: textDirection,
+              horizontalOffset: horizontalOffset,
+              verticalOffset: verticalOffset,
+              cells: cells,
+              cellSize: cellSize,
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class _FakeViewport extends SingleChildRenderObjectWidget {
+  const _FakeViewport({
+    Key key,
+    Widget child,
+  }) : super(key: key, child: child);
+
+  @override
+  _RenderFakeViewport createRenderObject(BuildContext context) => _RenderFakeViewport();
+}
+
+class _RenderFakeViewport extends RenderProxyBox implements RenderAbstractViewport {
+  _RenderFakeViewport({
+    RenderBox child,
+  }) : super(child);
+
+  @override
+  RevealedOffset getOffsetToReveal(RenderObject target, double alignment, {Rect rect}) {
+    // TODO(ianh): Implement this for real (and make these not be "Fake")
```

Comment:
oh, interesting. The API docs weren't clear about that.

Actually looks like what really matters is that the render object doesn't implement `showOnScreen`. I've added a dummy implementation for now. It's not actually clear to me how this object should work with accessibility. I mean, it probably needs a semantic label on each LatticeCell, and the accessibility logic should probably build a ton of accessibility nodes and the performance will suck... right now it's pretty much inaccessible since it's all just a big canvas with a few random nodes in it.

Reply:
Oh, right. The RenderViewportBase and the RenderSingleChildViewport are using getOffsetToReveal to implement showOnScreen, but you're not subclassing from those. So, yes, as you said, you'll need to implement showOnScreen instead (which I imagine would probably use getOffsetToReveal behind the scene, though).